File: src/main/kotlin/com/discord/oauth2rpc/structures/Presence.kt
package com.discord.oauth2rpc.structures

import com.discord.oauth2rpc.utils.*
import java.net.URI
import java.util.*

open class RichPresence(sessionId: String? = null) {

    var name: String? = null
    var type: Int = 0
    var url: String? = null
    var applicationId: String? = null
    var state: String? = null
    var details: String? = null
    var party: MutableMap<String, Any?>? = null
    var timestamps: MutableMap<String, Long?>? = null
    var buttons: MutableList<String> = mutableListOf()
    var platform: String? = null
    var secrets: MutableMap<String, Any?> = mutableMapOf()
    var metadata: MutableMap<String, Any?> = mutableMapOf()
    var createdTimestamp: Long = System.currentTimeMillis()
    var assets: MutableMap<String, String?> = mutableMapOf("largeImage" to null, "largeText" to null, "smallImage" to null, "smallText" to null)
    var flags: Int? = null
    var syncId: String? = null
    var id: String? = null
    var sessionId: String? = sessionId

    fun setApplicationId(id: String?): RichPresence { applicationId = id; return this }
    fun setType(type: Any): RichPresence {
        this.type = when (type) { is Int -> type; is String -> ActivityTypes.fromString(type) ?: 0; else -> 0 }
        return this
    }
    fun setURL(url: String?): RichPresence {
        if (url != null) { try { URI(url) } catch (_: Exception) { throw IllegalArgumentException("URL must be a valid URL") } }
        this.url = url; return this
    }
    fun setState(state: String?): RichPresence { this.state = state; return this }
    fun setName(name: String?): RichPresence { this.name = name; return this }
    fun setDetails(details: String?): RichPresence { this.details = details; return this }

    fun setParty(party: Map<String, Any?>?): RichPresence {
        if (party != null) {
            val maxVal = (party["max"] as? Number)?.toInt() ?: throw IllegalArgumentException("Party must have max number")
            val currentVal = (party["current"] as? Number)?.toInt() ?: throw IllegalArgumentException("Party must have current number")
            if (currentVal > maxVal) throw IllegalArgumentException("Party current must be less than max")
            this.party = mutableMapOf("size" to listOf(currentVal, maxVal), "id" to (party["id"] as? String ?: UUID.randomUUID().toString()))
        } else this.party = null
        return this
    }

    fun setStartTimestamp(timestamp: Any?): RichPresence {
        if (timestamps == null) timestamps = mutableMapOf("start" to null, "end" to null)
        timestamps!!["start"] = when (timestamp) { is Long -> timestamp; is Int -> timestamp.toLong(); is java.util.Date -> timestamp.time; else -> null }
        return this
    }

    fun setEndTimestamp(timestamp: Any?): RichPresence {
        if (timestamps == null) timestamps = mutableMapOf("start" to null, "end" to null)
        timestamps!!["end"] = when (timestamp) { is Long -> timestamp; is Int -> timestamp.toLong(); is java.util.Date -> timestamp.time; else -> null }
        return this
    }

    fun setAssetsLargeImage(image: String?): RichPresence { assets["largeImage"] = image; return this }
    fun setAssetsLargeText(text: String?): RichPresence { assets["largeText"] = text; return this }
    fun setAssetsSmallImage(image: String?): RichPresence { assets["smallImage"] = image; return this }
    fun setAssetsSmallText(text: String?): RichPresence { assets["smallText"] = text; return this }
    fun setPlatform(platform: String?): RichPresence { this.platform = platform; return this }

    fun addButton(name: String, url: String): RichPresence {
        require(name.isNotEmpty()) { "Button must have name" }; require(url.isNotEmpty()) { "Button must have url" }
        try { URI(url) } catch (_: Exception) { throw IllegalArgumentException("Button url must be a valid url") }
        buttons.add(name)
        @Suppress("UNCHECKED_CAST")
        (metadata.getOrPut("button_urls") { mutableListOf<String>() } as MutableList<String>).add(url)
        return this
    }

    fun setButtons(vararg button: Map<String, String>): RichPresence {
        if (button.isEmpty()) { buttons.clear(); metadata.remove("button_urls"); return this }
        if (button.size > 2) throw IllegalArgumentException("RichPresence can only have up to 2 buttons")
        buttons.clear(); metadata["button_urls"] = mutableListOf<String>()
        for (b in button) {
            val n = b["name"] ?: throw IllegalArgumentException("Button must have name")
            val u = b["url"] ?: throw IllegalArgumentException("Button must have url")
            try { URI(u) } catch (_: Exception) { throw IllegalArgumentException("Button url must be a valid url") }
            buttons.add(n)
            @Suppress("UNCHECKED_CAST")
            (metadata["button_urls"] as MutableList<String>).add(u)
        }
        return this
    }

    fun setJoinSecret(join: String?): RichPresence { secrets["join"] = join; return this }

    open fun toJSON(): Map<String, Any?> {
        val result = mutableMapOf<String, Any?>(
            "name" to name, "type" to type, "url" to url, "state" to state, "details" to details,
            "application_id" to applicationId,
            "timestamps" to timestamps?.let { mapOf("start" to it["start"], "end" to it["end"]) },
            "party" to party,
            "secrets" to if (secrets.isNotEmpty()) secrets.toMap() else null,
            "buttons" to if (buttons.isNotEmpty()) buttons.toList() else null,
            "metadata" to if (metadata.isNotEmpty()) metadata.toMap() else null,
            "platform" to platform, "created_at" to createdTimestamp, "session_id" to sessionId
        )
        if (flags != null) result["flags"] = flags
        if (syncId != null) result["sync_id"] = syncId
        if (id != null) result["id"] = id

        val hasAssets = assets["largeImage"] != null || assets["largeText"] != null || assets["smallImage"] != null || assets["smallText"] != null
        if (hasAssets) {
            result["assets"] = mapOf(
                "large_image" to Util.parseImage(assets["largeImage"]),
                "large_text" to assets["largeText"],
                "small_image" to Util.parseImage(assets["smallImage"]),
                "small_text" to assets["smallText"]
            )
        }
        return Util.clearNullOrUndefinedObject(result) ?: emptyMap()
    }
}

class CustomStatus {
    var name: String = " "
    var type: Int = ActivityTypes.fromString("CUSTOM") ?: 4
    var state: String? = null
    var emoji: Any? = null

    fun setEmoji(emoji: EmojiIdentifierResolvable?): CustomStatus { this.emoji = Util.resolvePartialEmoji(emoji); return this }
    fun setState(state: String?): CustomStatus {
        if (state != null && state.length > 128) throw IllegalArgumentException("State must be less than 128 characters")
        this.state = state; return this
    }
    fun toJSON(): Map<String, Any?> {
        if (emoji == null && state == null) throw IllegalStateException("CustomStatus must have at least one of emoji or state")
        return mapOf("name" to name, "type" to type, "state" to state, "emoji" to emoji)
    }
}

class SpotifyRPC(userId: String, sessionId: String? = null) : RichPresence(sessionId) {
    init {
        name = "Spotify"; type = ActivityTypes.fromString("LISTENING") ?: 2
        id = "spotify:1"; flags = 48
        party = mutableMapOf("id" to "spotify:$userId", "size" to emptyList<Int>())
    }

    fun setSongId(id: String): SpotifyRPC { syncId = id; return this }
    fun addArtistId(id: String): SpotifyRPC {
        @Suppress("UNCHECKED_CAST")
        val ids = (metadata.getOrPut("artist_ids") { mutableListOf<String>() } as MutableList<String>)
        ids.add(id); return this
    }
    fun setArtistIds(vararg ids: String): SpotifyRPC {
        metadata["artist_ids"] = ids.flatMap { listOf(it) }.toMutableList(); return this
    }
    fun setAlbumId(id: String): SpotifyRPC { metadata["album_id"] = id; metadata["context_uri"] = "spotify:album:$id"; return this }

    override fun toJSON(): Map<String, Any?> {
        val result = super.toJSON().toMutableMap()
        result.remove("id"); result.remove("emoji"); result.remove("platform"); result.remove("buttons")
        return result
    }
}

File: src/structures/Presence.ts
import { randomUUID } from 'node:crypto';
import { Util, EmojiIdentifierResolvable, ActivityTypes } from '../utils';

function parseImage(image: any): string | null {
	if (typeof image != 'string') {
		image = null;
	} else if (URL.canParse(image) && ['http:', 'https:'].includes(new URL(image).protocol)) {
		image = image
			.replace('https://cdn.discordapp.com/', 'mp:')
			.replace('http://cdn.discordapp.com/', 'mp:')
			.replace('https://media.discordapp.net/', 'mp:')
			.replace('http://media.discordapp.net/', 'mp:');
		if (!image.startsWith('mp:')) {
			throw new Error('INVALID_URL');
		}
	} else if (/^[0-9]{17,19}$/.test(image)) {
		// ID Assets
	} else if (['mp:', 'youtube:', 'spotify:', 'twitch:'].some((v) => image.startsWith(v))) {
		// Image
	} else if (image.startsWith('external/')) {
		image = `mp:${image}`;
	}
	return image;
}

export class RichPresence {
	public name?: string;
	public type: number = 0;
	public url?: string | null;
	public applicationId?: string | null;
	public state?: string | null;
	public details?: string | null;
	public party?: { id: string | null; size: number[] } | null;
	public timestamps?: { start: number | null; end: number | null } | null;
	public buttons: string[] = [];
	public platform?: string | null;
	public secrets: any = {};
	public metadata: any = {};
	public createdTimestamp: number = Date.now();
	public assets: {
		largeImage?: string | null;
		largeText?: string | null;
		smallImage?: string | null;
		smallText?: string | null;
	} = {};
	public flags?: number;
	public syncId?: string | null;
	public id?: string;
	public sessionId?: string | null;

	constructor(sessionId?: string) {
		if (sessionId) {
			this.sessionId = sessionId;
		}
	}

	public setApplicationId(id: string | null): this {
		this.applicationId = id;
		return this;
	}

	public setType(type: string | number): this {
		this.type = typeof type === 'number' ? type : (ActivityTypes[type] as number);
		return this;
	}

	public setURL(url: string | null): this {
		if (typeof url == 'string' && !URL.canParse(url)) throw new Error('URL must be a valid URL');
		this.url = url;
		return this;
	}

	public setState(state: string | null): this {
		this.state = state;
		return this;
	}

	public setName(name: string | null): this {
		this.name = name ?? undefined;
		return this;
	}

	public setDetails(details: string | null): this {
		this.details = details;
		return this;
	}

	public setParty(party: { max: number; current: number; id?: string } | null): this {
		if (party) {
			if (!party.max || typeof party.max != 'number') throw new Error('Party must have max number');
			if (!party.current || typeof party.current != 'number') throw new Error('Party must have current');
			if (party.current > party.max) throw new Error('Party current must be less than max number');
			this.party = {
				size: [party.current, party.max],
				id: party.id || randomUUID(),
			};
		} else {
			this.party = null;
		}
		return this;
	}

	public setStartTimestamp(timestamp: Date | number | null): this {
		if (!this.timestamps) this.timestamps = { start: null, end: null };
		if (timestamp instanceof Date) timestamp = timestamp.getTime();
		this.timestamps.start = timestamp;
		return this;
	}

	public setEndTimestamp(timestamp: Date | number | null): this {
		if (!this.timestamps) this.timestamps = { start: null, end: null };
		if (timestamp instanceof Date) timestamp = timestamp.getTime();
		this.timestamps.end = timestamp;
		return this;
	}

	public setAssetsLargeImage(image: string | null): this {
		this.assets.largeImage = image;
		return this;
	}

	public setAssetsLargeText(text: string | null): this {
		this.assets.largeText = text;
		return this;
	}

	public setAssetsSmallImage(image: string | null): this {
		this.assets.smallImage = image;
		return this;
	}

	public setAssetsSmallText(text: string | null): this {
		this.assets.smallText = text;
		return this;
	}

	public setPlatform(platform: string | null): this {
		this.platform = platform;
		return this;
	}

	public addButton(name: string, url: string): this {
		if (!name || !url) throw new Error('Button must have name and url');
		if (typeof name !== 'string') throw new Error('Button name must be a string');
		if (!URL.canParse(url)) throw new Error('Button url must be a valid url');
		this.buttons.push(name);
		if (Array.isArray(this.metadata.button_urls)) this.metadata.button_urls.push(url);
		else this.metadata.button_urls = [url];
		return this;
	}

	public setButtons(...button: { name: string; url: string }[]): this {
		if (button.length == 0) {
			this.buttons = [];
			delete this.metadata.button_urls;
			return this;
		} else if (button.length > 2) {
			throw new Error('RichPresence can only have up to 2 buttons');
		}

		this.buttons = [];
		this.metadata.button_urls = [];

		button.flat(2).forEach((b) => {
			if (b.name && b.url) {
				this.buttons.push(b.name);
				if (!URL.canParse(b.url)) throw new Error('Button url must be a valid url');
				this.metadata.button_urls.push(b.url);
			} else {
				throw new Error('Button must have name and url');
			}
		});
		return this;
	}

	public setJoinSecret(join: string | null): this {
		this.secrets.join = join;
		return this;
	}

	public toJSON(): any {
		const result: any = {
			name: this.name,
			type: this.type,
			url: this.url,
			state: this.state,
			details: this.details,
			application_id: this.applicationId,
			timestamps: this.timestamps,
			party: this.party,
			secrets: Object.keys(this.secrets).length > 0 ? this.secrets : undefined,
			buttons: this.buttons.length > 0 ? this.buttons : undefined,
			metadata: Object.keys(this.metadata).length > 0 ? this.metadata : undefined,
			platform: this.platform,
			created_at: this.createdTimestamp,
			session_id: this.sessionId,
		};

		if (this.flags !== undefined) result.flags = this.flags;
		if (this.syncId !== undefined) result.sync_id = this.syncId;
		if (this.id !== undefined) result.id = this.id;

		const hasAssets = this.assets.largeImage || this.assets.largeText || this.assets.smallImage || this.assets.smallText;
		if (hasAssets) {
			result.assets = {
				large_image: parseImage(this.assets.largeImage),
				large_text: this.assets.largeText,
				small_image: parseImage(this.assets.smallImage),
				small_text: this.assets.smallText,
			};
		}

		return Util.clearNullOrUndefinedObject(result);
	}
}

export class CustomStatus {
	public name: string = ' ';
	public type: number = ActivityTypes.CUSTOM as number;
	public state?: string | null;
	public emoji?: any | null;

	public setEmoji(emoji: EmojiIdentifierResolvable | null): this {
		this.emoji = Util.resolvePartialEmoji(emoji);
		return this;
	}

	public setState(state: string | null): this {
		if (typeof state == 'string' && state.length > 128) throw new Error('State must be less than 128 characters');
		this.state = state;
		return this;
	}

	public toJSON(): any {
		if (!this.emoji && !this.state) throw new Error('CustomStatus must have at least one of emoji or state');
		return {
			name: this.name,
			type: this.type,
			state: this.state,
			emoji: this.emoji,
		};
	}
}

export class SpotifyRPC extends RichPresence {
	constructor(userId: string, sessionId?: string) {
		super(sessionId);
		this.name = 'Spotify';
		this.type = ActivityTypes.LISTENING as number;
		this.id = 'spotify:1';
		this.flags = 48; // Sync + Play (ActivityFlags)
		this.party = {
			id: `spotify:${userId}`,
			size: [],
		} as any;
	}

	public setSongId(id: string): this {
		this.syncId = id;
		return this;
	}

	public addArtistId(id: string): this {
		if (!this.metadata.artist_ids) this.metadata.artist_ids = [];
		this.metadata.artist_ids.push(id);
		return this;
	}

	public setArtistIds(...ids: string[]): this {
		if (!ids?.length) {
			this.metadata.artist_ids = [];
			return this;
		}
		if (!this.metadata.artist_ids) this.metadata.artist_ids = [];
		ids.flat(2).forEach((id) => this.metadata.artist_ids.push(id));
		return this;
	}

	public setAlbumId(id: string): this {
		this.metadata.album_id = id;
		this.metadata.context_uri = `spotify:album:${id}`;
		return this;
	}

	public toJSON(): any {
		const result = super.toJSON();
		delete result.id;
		delete result.emoji;
		delete result.platform;
		delete result.buttons;
		return result;
	}
}
